mempool/src/core_mempool/index.rs

```diff
@@ -256,33 +257,81 @@ impl TimelineIndex {
 /// (because their sequence number is too high)
 /// we keep separate index to be able to efficiently evict them when Mempool is full
 pub struct ParkingLotIndex {
-    data: BTreeSet<TxnPointer>,
+    // DS invariants:
+    // 1. for each entry (account, txns) in `data`, `txns` is never empty
+    // 2. for all accounts, data.get(account_indices.get(`account`)) == (account, sequence numbers of account's txns)
+    data: Vec<(AccountAddress, BTreeSet<u64>)>,
+    account_indices: HashMap<AccountAddress, usize>,
+    size: usize,
 }
 
 impl ParkingLotIndex {
     pub(crate) fn new() -> Self {
         Self {
-            data: BTreeSet::new(),
+            data: vec![],
+            account_indices: HashMap::new(),
+            size: 0,
         }
     }
 
     /// add transaction to index
     pub(crate) fn insert(&mut self, txn: &MempoolTransaction) {
-        self.data.insert(TxnPointer::from(txn));
+        let sender = &txn.txn.sender();
+        let sequence_number = txn.txn.sequence_number();
+        let is_new_entry = match self.account_indices.get(sender) {
+            Some(index) => {
+                if let Some((_account, seq_nums)) = self.data.get_mut(*index) {
+                    seq_nums.insert(sequence_number)
+                } else {
+                    unreachable!("parking lot invariant violated: for account {}, account index exists but missing entry in data", sender);
+                }
+            }
+            None => {
+                let seq_nums = [sequence_number].iter().cloned().collect::<BTreeSet<_>>();
+                self.data.push((*sender, seq_nums));
+                self.account_indices.insert(*sender, self.data.len() - 1);
```

Comment:
Sorry, been too long since I've written code :/, self.account_indices.get(sender) is concurrency safe?  i.e. it locks access to the element?  Otherwise, you could reach the unreachable!() if you have two concurrent inserts

Reply:
I think all of core mempool is wrapped up in a mutex when accessed, so we don't have to worry about making the internal workings of core mempool concurrency-safe (in fact I'm p sure currently no aspect of core mempool's internal DS's are individually concurrency-safe)